File: algorithms/javaalgo/src/test/java/leetcode/number1to100/number31to40/number33/SearchInRotatedSortedArrayTest.java
package leetcode.number1to100.number31to40.number33;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

public class SearchInRotatedSortedArrayTest {

    @Test
    void test() {
        int[] nums = {4, 5, 6, 7, 0, 1, 2};
        int target = 0;
        int[] nums2 = {1, 3};
        int target2 = 3;

        assertThat(search(nums, target)).isEqualTo(4);
        assertThat(search(nums2, target2)).isEqualTo(1);
    }

    public int search(int[] nums, int target) {
        if (nums.length == 0) {
            return -1;
        }

        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        int pivot = left;
        left = 0;
        right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int midPivot = (mid + pivot) % nums.length;

            if (nums[midPivot] < target) {
                left = mid + 1;
            } else if (nums[midPivot] > target) {
                right = mid - 1;
            } else {
                return midPivot;
            }
        }
        return -1;
    }

    public int search2(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }

        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = (left + (right - left)) / 2;

            if (nums[mid] == target) {
                return mid;
            }

            if (nums[left] <= nums[mid]) {
                if (nums[left] <= target && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return -1;
    }
}

File: algorithms/javascriptalgo/leetcode/number1-100/number31-40/number33/searchInRotatedSortedArray.test.js
test("test", () => {
  const nums1 = [4, 5, 6, 7, 0, 1, 2];
  const target1 = 0;

  const nums2 = [4, 5, 6, 7, 0, 1, 2];
  const target2 = 3;

  const nums3 = [1];
  const target3 = 0;

  expect(search(nums1, target1)).toEqual(4);
  expect(search(nums2, target2)).toEqual(-1);
  expect(search(nums3, target3)).toEqual(-1);

});

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
const search = function (nums, target) {
  let left = 0;
  let right = nums.length - 1;

  while (left < right) {
    let mid = Math.floor(left + (right - left) / 2);
    if (nums[mid] > nums[right]) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }
  let pivot = left;
  left = 0;
  right = nums.length - 1;

  while (left <= right) {
    const mid = Math.floor(left + (right - left) / 2);
    const midPivot = (pivot + mid) % nums.length;
    if (nums[midPivot] === target) {
      return midPivot;
    } else if (nums[midPivot] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return -1;
};
